### PR TITLE
Fix netty dependency and class version checking

### PR DESCRIPTION
**Pull request by tylerbenson**
Still requires #417 to work all the way.

### CODE REVIEW

File: dd-java-agent/instrumentation/jax-rs-client/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsClientInstrumentation.java [hunk 1]
```diff
@@ -46,8 +46,10 @@ public ElementMatcher typeMatcher() {
   public static class ClientBuilderAdvice {
 
     @Advice.OnMethodEnter
-    public static void registerFeature(@Advice.This final ClientBuilder builder) {
-      builder.register(ClientTracingFeature.class);
+    public static void registerFeature(@Advice.This final Object builder) {
+      if (builder instanceof ClientBuilder) {
+        ((ClientBuilder) builder).register(ClientTracingFeature.class);
+      }
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/jax-rs-client/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsClientInstrumentation.java]: @tylerbenson could you explain why this is helpful: how instrumented class may not be present when its method is called?

Comment (gary-huang) [reply dd-java-agent/instrumentation/jax-rs-client/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsClientInstrumentation.java]: @tylerbenson and I were looking into this yesterday, and it seems that when `@Advice.This` is used with a specific class (i.e. `ClientBuilder` in this case), that class must to be already loaded at transformation time. Or ByteBuddy would throw the `Cannot resolve type description` error.

File: dd-java-agent/instrumentation/jms-2/src/main/java/datadog/trace/instrumentation/jms2/JMS2MessageProducerInstrumentation.java [hunk 2]
```diff
@@ -83,11 +83,11 @@ public static Scope startSpan(
           GlobalTracer.get()
```
Comment (gary-huang) [inline dd-java-agent/instrumentation/jms-2/src/main/java/datadog/trace/instrumentation/jms2/JMS2MessageProducerInstrumentation.java]: There's an `@Advice.This` here for the `MessageProducer` in `ProducerAdvice`'s `startSpan`, should this be an `Object`?

Comment (realark) [reply dd-java-agent/instrumentation/jms-2/src/main/java/datadog/trace/instrumentation/jms2/JMS2MessageProducerInstrumentation.java]: Would this be okay if the This param is an interface and the instrumentation applies to its implementations? Missing the full context here, but that type of pattern is okay for other types of advice/classloading issues.

File: dd-java-agent/instrumentation/jax-rs-client/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsClientInstrumentation.java (hunk 1 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/jax-rs-client/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsClientInstrumentation.java]: Thanks for explaining!
To me this seems like a problem on bytebuddy side (or on the way we run it) - and it makes our code much uglier than it should be in my opinion. Also as @ark mentions there are probably other places where same problem should arise but apparently doesn't. I hope there might be better way to fix this ran manually casting all advice's parameters :)

File: dd-java-agent/instrumentation/jms-2/src/main/java/datadog/trace/instrumentation/jms2/JMS2MessageProducerInstrumentation.java (hunk 2 above)
Comment (gary-huang) [reply dd-java-agent/instrumentation/jms-2/src/main/java/datadog/trace/instrumentation/jms2/JMS2MessageProducerInstrumentation.java]: That was the issue that we were seeing with Netty: https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/netty-4.1/src/main/java/datadog/trace/instrumentation/netty41/NettyChannelPipelineInstrumentation.java#L100

Where for the `@Advice.This` param, `ChannelPipeline` is an interface, and the implementation being instrumented is `DefaultChannelPipeline`.

**Comment by tylerbenson** [comment]
@realark I've removed the overlap with #417, but kept the part of this PR that fixes the dependency and version checking.  Should make this a much easier PR to review.

**Pull request merged**